File: app/src/main/java/com/worldventures/dreamtrips/wallet/service/command/wizard/ProcessSmartCardInfoDelegate.java
package com.worldventures.dreamtrips.wallet.service.command.wizard;

import com.worldventures.dreamtrips.api.smart_card.user_association.model.SmartCardInfo;
import com.worldventures.dreamtrips.wallet.domain.entity.ImmutableSmartCard;
import com.worldventures.dreamtrips.wallet.domain.entity.ImmutableSmartCardUser;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCard;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCardDetails;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCardUser;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCardUserPhone;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCardUserPhoto;
import com.worldventures.dreamtrips.wallet.domain.storage.WalletStorage;
import com.worldventures.dreamtrips.wallet.service.command.ActiveSmartCardCommand;

import io.techery.janet.Janet;
import io.techery.mappery.MapperyContext;
import rx.Observable;

class ProcessSmartCardInfoDelegate {

   private final WalletStorage walletStorage;
   private final Janet janetWallet;
   private MapperyContext mappery;

   ProcessSmartCardInfoDelegate(WalletStorage walletStorage, Janet janetWallet, MapperyContext mappery) {
      this.walletStorage = walletStorage;
      this.janetWallet = janetWallet;
      this.mappery = mappery;
   }

   Observable<Result> processSmartCardInfo(SmartCardInfo smartCardInfo) {
      final SmartCardDetails details = mappery.convert(smartCardInfo, SmartCardDetails.class);
      final SmartCard smartCard = createSmartCard(smartCardInfo);

      return createSmartCardUser(smartCardInfo.user())
            .map(user -> new Result(smartCard, user, details))
            .flatMap(result -> save(result).map(aVoid -> result));
   }

   private Observable<SmartCardUser> createSmartCardUser(com.worldventures.dreamtrips.api.smart_card.user_association.model.SmartCardUser user) {
      final ImmutableSmartCardUser.Builder userBuilder = ImmutableSmartCardUser.builder()
            .firstName(user.firstName() != null ? user.firstName() : "")
            .middleName(user.middleName() != null ? user.middleName() : "")
            .lastName(user.lastName() != null ? user.lastName() : "");

      if (user.phone() != null) userBuilder.phoneNumber(mappery.convert(user.phone(), SmartCardUserPhone.class));

      final SmartCardUser scUser = userBuilder.build();

      if (user.displayPhoto() != null) {
         final String photoUrl = user.displayPhoto();
         return Observable.just(photoUrl)
               .map(s -> changeUserPhoto(scUser, SmartCardUserPhoto.of(s)));
      } else {
         return Observable.just(scUser);
      }
   }

   private SmartCard createSmartCard(SmartCardInfo smartCardInfo) {
      return ImmutableSmartCard.builder()
            .smartCardId(String.valueOf(smartCardInfo.scId()))
            .deviceId(smartCardInfo.deviceId())
            .cardStatus(SmartCard.CardStatus.ACTIVE)
            .build();
   }

   private SmartCardUser changeUserPhoto(SmartCardUser user, SmartCardUserPhoto smartCardUserPhoto) {
      final String uri = smartCardUserPhoto.uri();
      if (uri != null) {
         return ImmutableSmartCardUser.builder()
               .from(user)
               .userPhoto(smartCardUserPhoto)
               .build();
      } else {
         return user;
      }
   }

   private Observable<Void> save(Result result) {
      walletStorage.saveSmartCardDetails(result.details);
      walletStorage.saveSmartCardUser(result.user);

      //saving smart card
      return janetWallet.createPipe(ActiveSmartCardCommand.class)
            .createObservableResult(new ActiveSmartCardCommand(result.smartCard))
            .map(command -> null);
   }

   static final class Result {
      public final SmartCard smartCard;
      public final SmartCardUser user;
      public final SmartCardDetails details;

      Result(SmartCard smartCard, SmartCardUser user, SmartCardDetails details) {
         this.smartCard = smartCard;
         this.user = user;
         this.details = details;
      }
   }
}

File: wallet/src/main/java/com/worldventures/wallet/service/command/wizard/ProcessSmartCardInfoDelegate.kt
package com.worldventures.wallet.service.command.wizard

import com.worldventures.core.utils.ProjectTextUtils.defaultIfEmpty
import com.worldventures.dreamtrips.api.smart_card.user_association.model.SmartCardInfo
import com.worldventures.wallet.domain.entity.ApiSmartCardUser
import com.worldventures.wallet.domain.entity.CardStatus
import com.worldventures.wallet.domain.entity.SmartCard
import com.worldventures.wallet.domain.entity.SmartCardDetails
import com.worldventures.wallet.domain.entity.SmartCardUser
import com.worldventures.wallet.domain.entity.SmartCardUserPhone
import com.worldventures.wallet.domain.entity.SmartCardUserPhoto
import com.worldventures.wallet.domain.storage.WalletStorage
import com.worldventures.wallet.service.command.ActiveSmartCardCommand
import io.techery.janet.Janet
import io.techery.mappery.MapperyContext
import rx.Observable

@Suppress("UnsafeCallOnNullableType")
internal class ProcessSmartCardInfoDelegate(
      private val walletStorage: WalletStorage,
      private val janetWallet: Janet,
      private val mappery: MapperyContext) {

   fun processSmartCardInfo(smartCardInfo: SmartCardInfo): Observable<Pair<SmartCard, SmartCardUser>> {
      val smartCard = createSmartCard(smartCardInfo)

      val user = createSmartCardUser(smartCardInfo.user())
      val result = Pair(smartCard, user)
      return save(result).map { result }
   }

   private fun createSmartCardUser(apiUser: ApiSmartCardUser): SmartCardUser {
      val photoUrl = apiUser.displayPhoto()
      val apiPhone = apiUser.phone()

      return SmartCardUser(
            firstName = defaultIfEmpty(apiUser.firstName(), ""),
            middleName = defaultIfEmpty(apiUser.middleName(), ""),
            lastName = defaultIfEmpty(apiUser.lastName(), ""),
            phoneNumber = if (apiPhone != null) mappery.convert(apiPhone, SmartCardUserPhone::class.java) else null,
            userPhoto = if (photoUrl.isNullOrEmpty()) null else SmartCardUserPhoto(photoUrl!!))
   }

   private fun createSmartCard(smartCardInfo: SmartCardInfo): SmartCard {
      return SmartCard(
            smartCardId = smartCardInfo.scId().toString(),
            cardStatus = CardStatus.ACTIVE,
            details = SmartCardDetails(
                  deviceId = smartCardInfo.deviceId(),
                  serialNumber = smartCardInfo.serialNumber(),
                  bleAddress = smartCardInfo.bleAddress(),
                  revVersion = smartCardInfo.revVersion(),
                  wvOrderId = smartCardInfo.wvOrderId(),
                  nxtOrderId = smartCardInfo.nxtOrderId(),
                  orderDate = smartCardInfo.orderDate())
      )
   }

   private fun save(result: Pair<SmartCard, SmartCardUser>): Observable<Void> {
      walletStorage.saveSmartCardUser(result.second)
      //saving smart card
      return janetWallet.createPipe(ActiveSmartCardCommand::class.java)
            .createObservableResult(ActiveSmartCardCommand(result.first))
            .map { null }
   }

}
